File: kotlin/2.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var v1 = l1?.`val` ?: 0
        var v2 = l2?.`val` ?: 0
        val ans = ListNode((v1 + v2) % 10)
        var aux = (v1 + v2) / 10

        var current = ans
        var p1 = l1
        var p2 = l2

        while (p1?.next != null || p2?.next != null) {
            p1 = p1?.next
            p2 = p2?.next
            
            v1 = p1?.`val` ?: 0
            v2 = p2?.`val` ?: 0

            val sum = v1 + v2 + aux
            aux = sum / 10

            current.next = ListNode(sum % 10)
            current = current.next
        }

        if (aux != 0) {
            current.next = ListNode(aux)
        }

        return ans
    }
}

File: typescript/2.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    let dummy: ListNode = new ListNode(0);
    let aux = 0
    let current = dummy

    let p1 = l1
    let p2 = l2

    while (p1 != null || p2 != null) {
        const val1 = p1?.val ?? 0
        const val2 = p2?.val ?? 0
        
        let sum = val1 + val2 + aux;

        current.next = new ListNode(sum % 10);
        current = current.next
        aux = Math.floor(sum / 10)

        p1 = p1?.next ?? null;
        p2 = p2?.next ?? null;
    }

    if (aux != 0) current.next = new ListNode(aux);

    return dummy.next;
};